src/order/upper_lower.lean

```diff
@@ -308,3 +338,81 @@ by simp_rw lower_set.compl_infi
 
 end lower_set
 end has_le
+
+/-! #### Principal sets -/
+
+namespace upper_set
+section preorder
+variables [preorder α] {a b : α}
+
+/-- The smallest upper set containing a given element. -/
+def principal (a : α) : upper_set α := ⟨{x | a ≤ x}, λ b c, ge_trans⟩
+
+@[simp] lemma coe_principal (a : α) : ↑(principal a) = {x | a ≤ x} := rfl
+@[simp] lemma mem_principal_iff : b ∈ principal a ↔ a ≤ b := iff.rfl
+
+end preorder
+
+section semilattice_sup
+variables [semilattice_sup α]
+
+@[simp] lemma principal_sup (a b : α) : principal (a ⊔ b) = principal a ⊓ principal b :=
+set_like.ext $ λ c, sup_le_iff
+
+end semilattice_sup
+
+section complete_lattice
+variables [complete_lattice α]
+
+@[simp] lemma principal_Sup (S : set α) : principal (Sup S) = ⨅ a ∈ S, principal a :=
+set_like.ext $ λ c, by simp only [mem_principal_iff, mem_infi₂_iff, Sup_le_iff]
+
+@[simp] lemma principal_supr (f : ι → α) : principal (⨆ i, f i) = ⨅ i, principal (f i) :=
+set_like.ext $ λ c, by simp only [mem_principal_iff, mem_infi_iff, supr_le_iff]
+
+@[simp] lemma principal_supr₂ (f : Π i, κ i → α) :
+  principal (⨆ i j, f i j) = ⨅ i j, principal (f i j) :=
+by simp_rw principal_supr
+
+end complete_lattice
+end upper_set
+
+namespace lower_set
+section preorder
+variables [preorder α] {a b : α}
+
+/-- The smallest lower set containing a given element. -/
+def principal (a : α) : lower_set α := ⟨{x | x ≤ a}, λ b c, le_trans⟩
+
+@[simp] lemma coe_principal (a : α) : ↑(principal a) = {x | x ≤ a} := rfl
+@[simp] lemma mem_principal_iff : b ∈ principal a ↔ b ≤ a := iff.rfl
+
+variables [order_top α]
+
+@[simp] lemma principal_top : principal (⊤ : α) = ⊤ := set_like.ext $ λ c, (iff_true _).2 le_top
+
+end preorder
+
+section semilattice_inf
+variables [semilattice_inf α]
+
+@[simp] lemma principal_inf (a b : α) : principal (a ⊓ b) = principal a ⊓ principal b :=
```

Comment:
This makes it look like `principal` is an `inf_hom`, and probably also an `inf_top_hom` and `Inf_hom`. Do you want bundled versions? And similar for the dual.

Reply:
The dual doesn't really work though, unless you want it to go from `order_dual α` to `upper_set α` (or unless we want to turn the order of `upper_set` over, which has the advantage that it matches `filter`, makes `compl` order-preserving, and gives a clear direction to the order on antichains).

Reply:
That's the usual way of having functions to/from the dual order, isn't it? Is there a reason it doesn't work here?

Reply:
Yeah, but do we really *care*?

Reply:
Are you suggesting that the bundled version is undesirable or unhelpful here? Do you think it's preferable to re-prove all of the results about bundled order homs for the new constructions here? If so, you should add them to this PR